src/core/Akka.Cluster/Cluster.cs

```diff
@@ -143,6 +143,17 @@ public Cluster(ActorSystemImpl system)
 
             system.RegisterOnTermination(Shutdown);
 
+            // Async join support
+            RegisterOnMemberUp(() =>
+            {
+                lock (_asyncJoinLock)
+                {
+                    _isUp.GetAndSet(true);
+                    // If there is an async join operation in progress, complete it.
+                    _asyncJoinTaskSource?.Complete();
```

Comment:
This looks a -little- closer...

IMO Ideally, we should only lock long enough to see if there is an existing continuation set up, and if so re-use that.

Does `_asyncJoinTaskSource` have a `TryComplete` to use instead?

Reply:
It uses `TryComplete` internally, the lock is to guard against possible null value

Reply:
`_asyncJoinTaskSource` is actually a custom wrapper class around `TaskCompletionSource` to provide timeout with custom exception support.